text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,801 @@
+
+
+
+## The Mangling Scheme by Example
+
+This section will develop an overview of the mangling scheme by walking through a number of examples. We'll start with the simplest case -- and will see how that already involves things that might be surprising.
+
+### Free-standing Functions and Statics
+
+A free-standing function is fully identified via its absolute path. For example, the following function
+
+```rust
+mod foo {
+  fn bar() {}
+}
+```
+
+has the path `foo::bar` and `N3foo3barE` is a mangling of that path that complies to the character set we are restricted to. Why this format with numbers embedded in it? It is the encoding that the [Itanium C++ ABI][itanium-mangling] name mangling scheme uses for "nested names" (i.e. paths). The scheme proposed here will also use this format because it does not need termination tokens for identifiers (which are hard to come by with our limited character set).
+
+However, the symbol name above does not unambiguously identify the function in every context. It is perfectly valid for another crate to also define `mod foo { fn bar() {} }` somewhere. So in order to avoid conflicts in such cases, the absolute path must always include the crate name and disambiguator, as in `N15mycrate_4a3b56d3foo3barE` (the crate disambiguator is used to disambiguate different versions of the same crate. It is an existing concept and not introduced by this RFC).
+
+There is another possible ambiguity that we have to take care of. Rust has two distinct namespaces: the type and the value namespace. This leads to a path of the form `crate_id::foo::bar` not uniquely identifying the item `bar` because the following snippet is legal Rust code:
+
+```rust
+fn foo() {
+    fn bar() {}
+}
+
+mod foo {
+    fn bar() {}
+}
+```
+
+The function `foo` lives in the value namespaces while the module `foo` lives in the type namespace. They don't interfere. In order to make the symbol names for the two distinct `bar` functions unique, we thus add a suffix to name components in the value namespace, so case one would get the symbol name `N15mycrate_4a3b56d3fooV3barVE` and case two get the name `N15mycrate_4a3b56d3foo3barVE` (notice the difference: `3fooV` vs `3foo`).
+
+There is on final case of name ambiguity that we have to take care of. Because of macro hygiene, multiple items with the same name can appear in the same context. The compiler internally disambiguates such names by augmenting them with a numeric index. For example, the first occurrence of the name `foo` within its parent is actually treated as `foo'0`, the second occurrence would be `foo'1`, the next `foo'2`, and so one. The mangling scheme will adopt this setup by appending a disambiguation suffix to each identifier with a non-zero index. So if macro expansion would result in the following code:
```

Comment:
```suggestion
There is one final case of name ambiguity that we have to take care of. Because of macro hygiene, multiple items with the same name can appear in the same context. The compiler internally disambiguates such names by augmenting them with a numeric index. For example, the first occurrence of the name `foo` within its parent is actually treated as `foo'0`, the second occurrence would be `foo'1`, the next `foo'2`, and so one. The mangling scheme will adopt this setup by appending a disambiguation suffix to each identifier with a non-zero index. So if macro expansion would result in the following code:
```